config/database.go

```diff
@@ -0,0 +1,322 @@
+	// Load the Postgres driver
+	_ "github.com/lib/pq"
+)
+
+// DatabaseStore is a config store backed by a database.
+type DatabaseStore struct {
+	emitter
+
+	config               *model.Config
+	environmentOverrides map[string]interface{}
+	configLock           sync.RWMutex
+	originalDsn          string
+	driverName           string
+	dataSourceName       string
+	db                   *sqlx.DB
+}
+
+// NewDatabaseStore creates a new instance of a config store backed by the given database.
+func NewDatabaseStore(dsn string) (ds *DatabaseStore, err error) {
+	driverName, dataSourceName, err := parseDSN(dsn)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid DSN")
+	}
+
+	// TODO: Retry logic?
+	db, err := sqlx.Open(driverName, dataSourceName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to connect to %s database", driverName)
+	}
+
+	ds = &DatabaseStore{
+		driverName:     driverName,
+		originalDsn:    dsn,
+		dataSourceName: dataSourceName,
+		db:             db,
+	}
+	if err = initializeConfigurationsTable(ds.db); err != nil {
+		return nil, errors.Wrap(err, "failed to initialize")
+	}
+
+	if err = ds.Load(); err != nil {
+		return nil, errors.Wrap(err, "failed to load")
+	}
+
+	return ds, nil
+}
+
+// parseDSN splits up a connection string into a driver name and data source name.
+//
+// For example:
+//	mysql://mmuser:mostest@dockerhost:5432/mattermost_test
+// returns
+//	driverName = mysql
+//	dataSourceName = mmuser:mostest@dockerhost:5432/mattermost_test
+//
+// By contrast, a Postgres DSN is returned unmodified.
+func parseDSN(dsn string) (string, string, error) {
+	// Treat the DSN as the URL that it is.
+	u, err := url.Parse(dsn)
+	if err != nil {
+		return "", "", errors.Wrap(err, "failed to parse DSN as URL")
+	}
+
+	scheme := u.Scheme
+	switch scheme {
+	case "mysql":
+		// Strip off the mysql:// for the dsn with which to connect.
+		u.Scheme = ""
+		dsn = strings.TrimPrefix(u.String(), "//")
+
+	case "postgres":
+		// No changes required
+
+	default:
+		return "", "", errors.Wrapf(err, "unsupported scheme %s", scheme)
+	}
+
+	return scheme, dsn, nil
+}
+
+// Get fetches the current, cached configuration.
+func (ds *DatabaseStore) Get() *model.Config {
+	ds.configLock.RLock()
+	defer ds.configLock.RUnlock()
+
+	return ds.config
+}
+
+// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+func (ds *DatabaseStore) GetEnvironmentOverrides() map[string]interface{} {
+	ds.configLock.RLock()
+	defer ds.configLock.RUnlock()
+
+	return ds.environmentOverrides
+}
+
+// Set replaces the current configuration in its entirety, without updating the backing store.
+func (ds *DatabaseStore) Set(newCfg *model.Config) (*model.Config, error) {
+	ds.configLock.Lock()
+	var unlockOnce sync.Once
+	defer unlockOnce.Do(ds.configLock.Unlock)
+
+	oldCfg := ds.config
+
+	// TODO: disallow attempting to save a directly modified config (comparing pointers). This
```

Comment:
This is the same TODO as the file store, but isn't going to be part of this change: it's tracked by separate JIRA issues to improve usage of the config store.